presto-hive/src/main/java/io/prestosql/plugin/hive/HiveBucketValidationRecordCursor.java

```diff
@@ -0,0 +1,129 @@
+    private final Path path;
+    private final int[] bucketColumnIndices;
+    private final List<Class<?>> javaTypeList;
+    private final List<TypeInfo> typeInfoList;
+    private final BucketingVersion bucketingVersion;
+    private final int bucketCount;
+    private final int expectedBucket;
+
+    private final Object[] scratch;
+
+    public HiveBucketValidationRecordCursor(
+            Path path,
+            int[] bucketColumnIndices,
+            List<HiveType> bucketColumnTypes,
+            BucketingVersion bucketingVersion,
+            int bucketCount,
+            int expectedBucket,
+            TypeManager typeManager,
+            RecordCursor delegate)
+    {
+        this.path = requireNonNull(path, "path is null");
+        this.bucketColumnIndices = requireNonNull(bucketColumnIndices, "bucketColumnIndices is null");
+        requireNonNull(bucketColumnTypes, "bucketColumnTypes is null");
+        this.javaTypeList = bucketColumnTypes.stream()
+                .map(HiveType::getTypeSignature)
+                .map(typeManager::getType)
+                .map(Type::getJavaType)
+                .collect(toImmutableList());
+        this.typeInfoList = bucketColumnTypes.stream()
+                .map(HiveType::getTypeInfo)
+                .collect(toImmutableList());
+        this.bucketingVersion = requireNonNull(bucketingVersion, "bucketingVersion is null");
+        this.bucketCount = bucketCount;
+        this.expectedBucket = expectedBucket;
+        this.delegate = requireNonNull(delegate, "delegate is null");
+
+        this.scratch = new Object[bucketColumnTypes.size()];
+    }
+
+    @VisibleForTesting
+    @Override
+    public RecordCursor delegate()
+    {
+        return delegate;
+    }
+
+    @Override
+    public boolean advanceNextPosition()
+    {
+        if (!delegate.advanceNextPosition()) {
+            return false;
+        }
+
+        for (int i = 0; i < scratch.length; i++) {
+            int index = bucketColumnIndices[i];
+            if (delegate.isNull(index)) {
+                scratch[i] = null;
+                continue;
+            }
+            Class<?> javaType = javaTypeList.get(i);
+            if (javaType == boolean.class) {
+                scratch[i] = delegate.getBoolean(index);
+            }
+            else if (javaType == long.class) {
+                scratch[i] = delegate.getLong(index);
+            }
+            else if (javaType == double.class) {
+                scratch[i] = delegate.getDouble(index);
+            }
+            else if (javaType == Slice.class) {
+                scratch[i] = delegate.getSlice(index);
+            }
+            else if (javaType == Block.class) {
+                scratch[i] = delegate.getObject(index);
+            }
+            else {
+                throw new VerifyException("Unknown Java type: " + javaType);
+            }
+        }
+
+        int bucket = getHiveBucket(bucketingVersion, bucketCount, typeInfoList, scratch);
```

Comment:
Should i be concerned about overhead of this?
would it be enough to do this very other row?
or eg first 100 rows, and then every 10th row?